File: Lab 10/t1.go
package main

import (
  "fmt"
  "os"
  "bufio"
  "strings"
  "strconv"
)

func main() {
  os.Exit(t1_main())
}

func t1_main() int {
  for {
    fmt.Println("Please input a binary number:")
    line, e := bufio.NewReader(os.Stdin).ReadString('\n')
    if e != nil {
      fmt.Fprintln(os.Stderr, "Read error:", e)
      return 1
    }
    // NOTE: ParseInt will refuse to convert a string with any illegal
    // characters.
    line = strings.TrimSuffix(line, "\n")

    num, e := strconv.ParseInt(line, 2, 0)
    if e != nil {
      fmt.Fprintln(os.Stderr, "Convert error:", e)
      return 1
    }

    fmt.Printf("%b is %d in decimal.\n", num, num)
  }
}

File: Lab 10/t1.rs
use std::io::{stdin, BufRead};

fn main() {
  loop {
    println!("Please input a binary number:");

    // NOTE: This handle is required to live longer than its lock() method in
    // order to allow "borrowing".
    let stdin = stdin();

    let buf = match stdin.lock().lines().next() {
      Some(r) => match r {
        // NOTE: Empty string is not None.
        Ok(r) => if r.len() > 0 { r } else { return; },
        Err(e) => {
          eprintln!("Read error: {}", e);
          return;
        }
      },
      None => {
        return;
      }
    };

    let num = match u32::from_str_radix(buf.as_str(), 2) {
      Ok(r) => r,
      Err(e) => {
        eprintln!("Parse error: {}", e);
        return;
      }
    };

    println!("{:b} is {} in decimal.", num, num);
  }
}
